File: py/wk1/binary_search.py
# @author { @Override } | @date 20250318 @ 14:15


def binarySearch(array: list[int], target: int) -> int:
    low: int = 0
    high: int = len(array) - 1
    while low <= high:
        m: int = low + (high - low) // 2
        if array[m] == target:
            return m
        elif array[m] < target:
            low = m + 1
        else:
            high = m - 1
    return -1  # Target not found


def main():

    # Won't work because it is NOT IN ORDER!
    myList: list[int] = [55, 80, 69, 88, 15, 66]
    print(f"Array before search: {myList}")
    print(f"Key searched: {binarySearch(myList, 88)}")

    # Will work because is in order
    sortedList: list[int] = [15, 25, 35, 45, 55, 65, 99, 101, 155]
    print(f"Array before search: {sortedList}")
    print(f"index[{binarySearch(sortedList, 55)}]")


main()

File: ocaml/wk1/BinarySearch.ml
(* @author: { @Override } | @date 20250320 : 1202*)

let bSearch(arr, target: int): int =
  let low: int = 0;
  let high: int = Array.length arr - 1;
  while low <= high do
    let mid: int = (low + high) / 2;
    if arr.[mid] = target then
      mid  (* Target found, return index *)
    else if arr.[mid] < target then
      low <- mid + 1  (* Search in the right half *)
    else
      high <- mid - 1  (* Search in the left half *)


let () =
  let arr = [|1; 2; 3; 4; 5; 6; 7; 8; 9; 10|] in
  let target = 7 in
  let result = bSearch(arr, target) in
  if result >= 0 then
    Printf.printf "Element %d found at index %d\n" target result
  else
    Printf.printf "Element %d not found in the array\n" target



